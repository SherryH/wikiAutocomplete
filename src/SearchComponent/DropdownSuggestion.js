import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

const DropdownSuggestion = ({
  classes, dropdownData, toggledClass, selectDropdown,
}) => (
  <ul className={classNames(toggledClass, { [classes.dropdownWrapper]: true })}>
    {dropdownData.map(dropdownItem => (
      <li key={dropdownItem} onClick={() => selectDropdown(dropdownItem)}>
        {dropdownItem}
      </li>
    ))}
  </ul>
);

export default DropdownSuggestion;

DropdownSuggestion.propTypes = {
  selectDropdown: PropTypes.func,
  toggledClass: PropTypes.object,
  classes: PropTypes.any,
  dropdownData: PropTypes.arrayOf(PropTypes.string),
};
