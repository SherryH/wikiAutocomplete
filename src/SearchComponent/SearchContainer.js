import React from 'react';
import PropTypes from 'prop-types';
import jsonp from 'jsonp';
import { from, Observable } from 'rxjs';
import {
  debounceTime, distinctUntilChanged, map, filter, switchMap,
} from 'rxjs/operators';

class SearchContainer extends React.Component {
  constructor(props) {
    super(props);
    this.searchInput$ = null;
    this.searchInputSubscriber = null;
    this.wikiSubscriber = null;
    this.searchInputRef = null;
  }

  state = {
    isOpen: false,
    searchValue: '',
    dropdownData: [],
    searchInput: null,
  };


  componentDidUpdate() {
    const { searchValue } = this.state;
    console.log('componentDidUpdate in test? ', this.searchInput$);
    if (!this.searchInput$) return;
    this.searchInputSubscriber = this.searchInput$
      .pipe(
        debounceTime(500),
        distinctUntilChanged(),
        filter(() => searchValue.length > 0),
        switchMap(() => {
          this.wikiSubscriber = this.getWikiObservable(searchValue);
          return this.wikiSubscriber;
        }),
      )
      .subscribe((dropdownData) => {
        // TODO: why from(promise) or(promise) doesnt work?
        // from(this.searchWiki(this.state.searchValue))
        //   .pipe(mergeAll())
        //   .subscribe((data) => {
        //     console.log('returned wiki data', data);
        //   });
        console.log('TODO: observable subscribe run increments', this.state.searchValue);

        this.setState({ dropdownData });
      });
  }

  componentWillUnmount() {
    if (this.searchInputSubscriber) this.searchInputSubscriber.unsubscribe();
    if (this.wikiSubscriber) this.wikiSubscriber.unsubscribe();
  }

  onClick = () => {
    this.setState(({ isOpen }) => ({
      isOpen: !isOpen,
    }));
  };


  getJsonpAsync = (term, url) => new Promise((resolve, reject) => {
    jsonp(url, (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data[1]);
    });
  });


  getWikiObservable = (term) => {
    let cancelled = false;
    return Observable.create((observer) => {
      const url = `https://en.wikipedia.org/w/api.php?action=opensearch&format=json&formatversion=2&search=${term}&namespace=0&limit=10&suggest=true`;
      const jsonPFunc = function (error, jsonpData) {
        if (error) observer.error(error);
        if (!cancelled) {
          observer.next(jsonpData[1]);
          observer.complete();
        }
      };
      jsonp(url, jsonPFunc);

      return function dispose() {
        cancelled = true;
      };
    });
  };

  onChange = (event, searchInput$) => {
    if (!this.searchInput$) {
      this.searchInput$ = searchInput$;
    }
    this.setState({ searchValue: event.target.value }, () => {
    });
  };

  selectDropdown = (dropdownList) => {
    this.setState({ searchValue: dropdownList, dropdownData: [] });
    // TODO: cursor still not displayed
    if (this.searchInputRef) this.searchInputRef.current.focus();
  }

  getSearchInput = (searchInputRef) => {
    // TODO: find the best place to define searchInputRef
    // passed through in onChange or another place?
    // Note: constructor() doesnt work as SearchBox not yet exist when SearchContainer defined
  };


  getStateAndHelpers() {
    const { isOpen, searchValue, dropdownData } = this.state;
    return {
      isOpen,
      onClick: this.onClick,
      handleClickOutside: this.handleClickOutside,
      searchWiki: this.searchWiki,
      onChange: this.onChange,
      searchValue,
      getSearchInput: this.getSearchInput,
      dropdownData,
      selectDropdown: this.selectDropdown,
    };
  }

  searchWiki = (term) => {
    const url = `https://en.wikipedia.org/w/index.php?search=${encodeURI(term)}`;
    window.open(url, '_blank');
  };

  handleClickOutside = () => {
    this.setState({
      isOpen: false,
    });
  };

  render() {
    const { children } = this.props;

    return children(this.getStateAndHelpers());
  }
}

export default SearchContainer;

SearchContainer.propTypes = {
  children: PropTypes.func.isRequired,
};
