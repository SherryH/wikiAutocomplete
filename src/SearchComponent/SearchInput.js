import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Subject } from 'rxjs';
import {
  debounceTime, distinctUntilChanged,
} from 'rxjs/operators';
import jsonp from 'jsonp';

export const input$ = new Subject();

class SearchInput extends React.PureComponent {
  state = {
    searchValue: '',
  };

  componentDidMount() {
    const { setDropdownData } = this.props;
    input$
      .pipe(
        distinctUntilChanged(),
        debounceTime(500),
      )
      .subscribe((searchValue) => {
        this.getJsonpAsync(searchValue).then((dropdownData) => {
          setDropdownData(dropdownData);
        });
      });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.searchValue !== this.props.searchValue) {
      this.setState({ searchValue: this.props.searchValue });
    }
  }

  handleChange = (event) => {
    this.setState({ searchValue: event.target.value }, () => {
      const { searchValue } = this.state;
      input$.next(searchValue);
    });
  };

  getJsonpAsync = term => new Promise((resolve, reject) => {
    const url = `https://en.wikipedia.org/w/api.php?action=opensearch&format=json&formatversion=2&search=${term}&namespace=0&limit=10&suggest=true`;
    jsonp(url, (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data[1]);
    });
  });

  render() {
    const { toggledClass } = this.props;
    const { searchValue } = this.state;
    return (
      <input
        data-testid="searchInput"
        id="searchInput"
        onChange={this.handleChange}
        value={searchValue}
        className={classNames(toggledClass)}
        placeholder="Start Wiki Search..."
      />
    );
  }
}

export default SearchInput;

SearchInput.propTypes = {
  toggledClass: PropTypes.object,
  setDropdownData: PropTypes.func,
  searchValue: PropTypes.string,
};

SearchInput.defaultProps = {
  toggledClass: {},
  setDropdownData: () => {},
  searchValue: '',
};
