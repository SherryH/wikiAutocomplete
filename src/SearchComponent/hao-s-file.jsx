
import {Subject}from 'rxjs'
import { debounceTime } from 'rxjs/operators';

const input$=new Subject()

class Input extends React.PureComponent{

    componentDidMount(){
input$.pipe(
    debounceTime(500)
).subscribe(
    this.callApi()
)

callApi(){
    return fetch('..')
    .then(stuff=>{
        this.setState(stuff)
    })
}

    }
    render(){
        return <input value={this.state.value} onChange={()=>
            this.setState({input},input$.onNext)
        }/>
    }
}