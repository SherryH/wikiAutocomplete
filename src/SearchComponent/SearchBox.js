import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Search } from '@material-ui/icons';
import { IconButton } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import onhandleClickOutside from 'react-onclickoutside';
import DropdownSuggestion from './DropdownSuggestion';
import SearchInput from './SearchInput';

const primaryBlue = '#1eaddc';

const useStyles = withStyles({
  open: {
    flex: 1,
    fontSize: 20,
    border: 0,
    '&:focus': {
      outline: 'none',
    },
  },
  closed: {
    display: 'none',
  },
  show: {
    display: 'block',
  },
  icon: {
    width: 40,
    height: 40,
    '&.open': {
      fill: primaryBlue,
    },
  },
  wrapper: {
    border: `1px solid ${primaryBlue}`,
    width: 65,
    borderRadius: '50%',
    background: primaryBlue,
    boxShadow: '0 10px 30px #d0d0d0',
    transition: '0.15s ease width',
    '&.open': {
      display: 'flex',
      borderRadius: '20%/110%',
      width: 380,
      background: 'white',
      padding: '2px 30px',
      position: 'relative',
    },
  },
  dropdownWrapper: {
    position: 'absolute',
    top: '68px',
    listStyle: 'none',
    border: '1px solid #999',
    width: 'calc(350px)',
    marginTop: '0',
    padding: '4px',
    maxHeight: '143px',
    overflowY: 'auto',
    background: 'white',
    borderRadius: '0 0 5px 5px',
    boxShadow: '0 10px 30px #999',
    '& li': {
      padding: '0.3rem',
      borderBottom: '1px solid #eee',
      background: 'white',
      '&:last-child': {
        borderBottom: 0,
      },
      '&:hover': {
        backgroundColor: primaryBlue,
        color: 'white',
        cursor: 'pointer',
        fontWeight: 700,
      },
    },
  },
});

const searchWiki = (term) => {
  const url = `https://en.wikipedia.org/w/index.php?search=${encodeURI(term)}`;
  window.open(url, '_blank');
};

const SearchBox = (props) => {
  const {
    classes,
    isOpen,
    onClick,
    searchValue,
    dropdownData,
    setDropdownData,
    selectDropdown,
  } = props;

  const toggledClass = { [classes.open]: isOpen, [classes.closed]: !isOpen };

  return (
    <React.Fragment>
      <form
        className={classNames(classes.wrapper, { open: isOpen })}
        onSubmit={() => searchWiki(searchValue)}
        autoComplete="off"
      >
        <IconButton
          data-testid="show-autocomplete"
          aria-label="Show autocomplete"
          onClick={onClick}
          className={classNames({ [classes.show]: !isOpen, [classes.closed]: isOpen })}
        >
          <Search className={classNames(classes.icon, { open: isOpen })} />
        </IconButton>
        <SearchInput
          searchValue={searchValue}
          toggledClass={toggledClass}
          setDropdownData={setDropdownData}
        />
        {dropdownData && dropdownData.length > 0 && (
          <DropdownSuggestion
            classes={classes}
            dropdownData={dropdownData}
            toggledClass={toggledClass}
            selectDropdown={selectDropdown}
          />
        )}
        <IconButton
          aria-label="Search"
          onClick={() => searchWiki(searchValue)}
          className={classNames({ [classes.show]: isOpen, [classes.closed]: !isOpen })}
        >
          <Search className={classNames(classes.icon, { open: isOpen })} />
        </IconButton>
      </form>
    </React.Fragment>
  );
};
export default onhandleClickOutside(useStyles(SearchBox), {
  handlehandleClickOutside(instance) {
    return () => {
      instance.props.handleClickOutside();
    };
  },
});

SearchBox.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  setDropdownData: PropTypes.func.isRequired,
  selectDropdown: PropTypes.func.isRequired,
  searchValue: PropTypes.string.isRequired,
  handleClickOutside: PropTypes.func.isRequired,
  classes: PropTypes.any,
  dropdownData: PropTypes.arrayOf(PropTypes.string),
};
