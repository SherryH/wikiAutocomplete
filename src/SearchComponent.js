import React from 'react';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import SearchBox from './SearchBox';
import SearchContainer from './SearchContainer';

const SearchComponent = ({ classes }) => (
  <SearchContainer>
    {({ isOpen, onClick, clickOutside }) => (
      <React.Fragment>
        <div className={classNames(classes.root, { open: isOpen })}>
          <SearchBox isOpen={isOpen} onClick={onClick} clickOutside={clickOutside} />
        </div>
      </React.Fragment>
    )}
  </SearchContainer>
);

const useStyles = withStyles({
  root: {
    width: '100vw',
    height: '100vh',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    transition: '2s ease all',
    '&.open': {
      background: '#1eaddc',
    },
  },
});

export default useStyles(SearchComponent);
