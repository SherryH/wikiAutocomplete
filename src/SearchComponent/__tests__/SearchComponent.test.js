jest.mock('jsonp');

import React from 'react';
import renderer from 'react-test-renderer';
import {
  render, fireEvent, cleanup, wait
} from 'react-testing-library';
// import jsonp from 'jsonp';
import SearchComponent from '../index';


afterAll(() => {
  jest.unmock('jsonp');
});

describe('SearchComponent:', () => {
  xit('closed autocomplete input renders correctly', () => {
    const tree = renderer
      .create(<SearchComponent />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  xit('opened autocomplete input renders correctly', () => {
    // render the autocomplete
    const { getByTestId, container } = render(
      <SearchComponent />,
    );
    // click the button to open the autocomplete
    fireEvent.click(getByTestId('show-autocomplete'));
    // snapshot testing to match the opened state
    expect(container.firstChild).toMatchSnapshot();
    cleanup();
  });

  it('renders search results into dropdown suggestions', async() => {
    // mock the wiki result returns by mocking jsonp

    // render the autocomplete
    const { getByTestId, container } = render(
      <SearchComponent />,
    );

    // input search text into the input field
    const searchInput = getByTestId('searchInput');
    // searchInput.value = 'London';
    fireEvent.change(searchInput, { target: { value: 'London' } });

    // check mock jsonp called
    // currently failed. refer https://github.com/kentcdodds/react-testing-library/blob/cb14d2d3a19ae1420fafc8d9554c707fd7123a99/src/__tests__/fetch.js
    // wait for componentDidUpdate called
    // expect(jsonp).toHaveBeenCalled();
    await wait()

    // match snapshots
    expect(container.firstChild).toMatchSnapshot();

    // check dropdownSelection is rendered

    cleanup();
  });

  it('opens a separate tab for the selected result', () => {
    // render the autocomplete
    const { getByTestId, container } = render(
      <SearchComponent />,
    );

    // input search text into the input field

    // mock the wiki result returns

    // click on one result

    // click on search

    // assert url

    cleanup();
  });
});
