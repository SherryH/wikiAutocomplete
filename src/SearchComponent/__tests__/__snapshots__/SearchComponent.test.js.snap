// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent: renders search results into dropdown suggestions 1`] = `
<div
  class="SearchComponent-root-1"
>
  <form
    autocomplete="off"
    class="SearchBox-wrapper-6"
  >
    <button
      aria-label="Show autocomplete"
      class="MuiButtonBase-root-14 MuiIconButton-root-8 SearchBox-show-4"
      data-testid="show-autocomplete"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-13"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-17 SearchBox-icon-5"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root-26"
      />
    </button>
    <input
      class="SearchBox-closed-3"
      data-testid="searchInput"
      id="searchInput"
      placeholder="Start Wiki Search..."
      value="London"
    />
    <button
      aria-label="Search"
      class="MuiButtonBase-root-14 MuiIconButton-root-8 SearchBox-closed-3"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-13"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-17 SearchBox-icon-5"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root-26"
      />
    </button>
  </form>
</div>
`;
